presto-main/src/main/java/io/prestosql/connector/informationschema/InformationSchemaMetadata.java

```diff
@@ -247,97 +240,56 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+        //   Discrete and range schema-level filters are supported;
+        //   discrete tabel-level filters are supported;
+        //   range table-level filters are not yet supported.
+        Optional<Set<String>> newSchemaNames = filterString(constraint.getSummary(), SCHEMA_COLUMN_HANDLE);
+        if (!newSchemaNames.isPresent()) {
+            if (table.getSchemaNames().isPresent()) {
+                newSchemaNames = table.getSchemaNames();
+            }
+            else {
+                newSchemaNames = Optional.of(enumerateSchemaNames(metadata, session, catalogName));
+            }
+        }
+        if (constraint.predicate().isPresent()) {
+            newSchemaNames = Optional.of(newSchemaNames.get().stream()
+                    .filter(schemaName -> constraint.predicate().get().test(schemaAsFixedValues(schemaName)))
+                    .collect(toImmutableSet()));
+        }
 
-    private Set<QualifiedTablePrefix> getPrefixes(ConnectorSession session, InformationSchemaTableHandle table, Constraint constraint)
-    {
-        InformationSchemaTableHandle handle = checkTableHandle(table);
+        Optional<Set<String>> newTableNames = filterString(constraint.getSummary(), TABLE_NAME_COLUMN_HANDLE);
 
-        if (constraint.getSummary().isNone()) {
-            return ImmutableSet.of();
+        // Neither the schema prefixes nor the table name domain changes.
+        if ((table.getSchemaNames().isPresent() && newSchemaNames.get().size() == table.getSchemaNames().get().size())
+                && (!newTableNames.isPresent() || (table.getTableNames().isPresent() && table.getTableNames().get().size() == newTableNames.get().size()))) {
+            return Optional.empty();
         }
 
-        Set<QualifiedTablePrefix> prefixes = calculatePrefixesWithSchemaName(session, constraint.getSummary(), constraint.predicate());
-        if (isTablesEnumeratingTable(handle.getSchemaTableName())) {
-            Set<QualifiedTablePrefix> tablePrefixes = calculatePrefixesWithTableName(session, prefixes, constraint.getSummary(), constraint.predicate());
-            // in case of high number of prefixes it is better to populate all data and then filter
-            if (tablePrefixes.size() <= MAX_PREFIXES_COUNT) {
-                prefixes = tablePrefixes;
-            }
-        }
-        if (prefixes.size() > MAX_PREFIXES_COUNT) {
-            // in case of high number of prefixes it is better to populate all data and then filter
-            prefixes = defaultPrefixes();
+        if (!newTableNames.isPresent() && table.getTableNames().isPresent()) {
+            newTableNames = table.getTableNames();
         }
 
-        return prefixes;
+        table = new InformationSchemaTableHandle(table.getCatalogName(), table.getSchemaName(), table.getTableName(), newSchemaNames, newTableNames);
```

Comment:
I don't see a pleace where you combine old predicates with new predicates (`newSchemaNames` with `table.getSchemaNames()`). And here you should pass `resultSchemaNames` and `resultTableNames`